File: backend/app/services/chess_com.py
"""Chess.com Published-Data API client.

The public API needs no authentication, but it *does* reject requests without a
descriptive User-Agent, so every call goes through `_client()`.
"""

from __future__ import annotations

import io
import logging
from datetime import datetime, timezone

import chess.pgn
import httpx

from app.config import settings

log = logging.getLogger(__name__)

BASE = "https://api.chess.com/pub"

DRAW_RESULTS = {
    "agreed",
    "repetition",
    "stalemate",
    "insufficient",
    "50move",
    "timevsinsufficient",
}


class ChessComError(RuntimeError):
    pass


def _client() -> httpx.Client:
    return httpx.Client(
        timeout=20.0,
        follow_redirects=True,
        headers={"User-Agent": settings.chess_com_user_agent, "Accept": "application/json"},
    )


def player_exists(username: str) -> bool:
    with _client() as c:
        r = c.get(f"{BASE}/player/{username.lower()}")
        return r.status_code == 200


def list_archives(username: str) -> list[str]:
    """Monthly archive URLs, oldest first."""
    with _client() as c:
        r = c.get(f"{BASE}/player/{username.lower()}/games/archives")
        if r.status_code == 404:
            raise ChessComError(f"Chess.com user '{username}' not found")
        r.raise_for_status()
        return r.json().get("archives", [])


def fetch_archive(url: str) -> list[dict]:
    with _client() as c:
        r = c.get(url)
        if r.status_code == 404:
            return []
        r.raise_for_status()
        return r.json().get("games", [])


def fetch_current_month(username: str) -> list[dict]:
    now = datetime.now(timezone.utc)
    return fetch_archive(f"{BASE}/player/{username.lower()}/games/{now.year}/{now.month:02d}")


def fetch_recent_months(username: str, months: int) -> list[dict]:
    """Last `months` archives (including the current one), oldest first."""
    archives = list_archives(username)
    games: list[dict] = []
    for url in archives[-max(months, 1) :]:
        games.extend(fetch_archive(url))
    return games


def _pgn_headers(pgn: str) -> dict:
    try:
        game = chess.pgn.read_game(io.StringIO(pgn))
    except Exception:  # malformed PGN should never kill a sync
        return {}
    return dict(game.headers) if game else {}


def normalize_game(raw: dict, chess_com_username: str) -> dict | None:
    """Turn a Chess.com archive entry into kwargs for `Game`.

    Returns None for anything we cannot analyse (variants, missing PGN).
    """
    if raw.get("rules") != "chess":
        return None
    pgn = raw.get("pgn")
    if not pgn:
        return None

    me = chess_com_username.lower()
    white = raw.get("white") or {}
    black = raw.get("black") or {}

    if (white.get("username") or "").lower() == me:
        mine, theirs, color = white, black, "white"
    elif (black.get("username") or "").lower() == me:
        mine, theirs, color = black, white, "black"
    else:
        return None

    raw_result = mine.get("result", "")
    if raw_result == "win":
        result = "win"
    elif raw_result in DRAW_RESULTS:
        result = "draw"
    else:
        result = "loss"

    # Termination is whichever side's result code is not the generic "win".
    termination = theirs.get("result") if raw_result == "win" else raw_result

    headers = _pgn_headers(pgn)
    end_time = int(raw.get("end_time") or 0)

    # Only present once the game has been reviewed on Chess.com.
    accuracies = raw.get("accuracies") or {}
    chess_com_accuracy = accuracies.get(color)

    return {
        "chess_com_game_id": str(raw.get("uuid") or raw.get("url") or end_time),
        "url": raw.get("url"),
        "pgn": pgn,
        "user_color": color,
        "user_rating": mine.get("rating"),
        "opponent_username": theirs.get("username") or "?",
        "opponent_rating": theirs.get("rating"),
        "result": result,
        "termination": termination,
        "time_class": raw.get("time_class"),
        "time_control": raw.get("time_control"),
        "rated": bool(raw.get("rated", True)),
        "eco": (headers.get("ECO") or None),
        "opening": _opening_name(headers),
        "chess_com_accuracy": chess_com_accuracy,
        "played_at": datetime.fromtimestamp(end_time, tz=timezone.utc),
        "end_time": end_time,
    }


def _opening_name(headers: dict) -> str | None:
    """Chess.com puts the opening in ECOUrl, e.g. .../openings/Sicilian-Defense-Najdorf."""
    if headers.get("Opening"):
        return headers["Opening"][:160]
    url = headers.get("ECOUrl")
    if not url:
        return None
    return url.rstrip("/").rsplit("/", 1)[-1].replace("-", " ")[:160]

File: frontend/src/data/chessCom.js
/**
 * Chess.com Published-Data API client, from backend/app/services/chess_com.py.
 *
 * The API is public and unauthenticated but rejects requests without a
 * descriptive User-Agent, so every call goes through `request`.
 *
 * Requests go out through CapacitorHttp, which performs them on the native HTTP
 * stack rather than in the WebView. That is what makes this app possible
 * without a server: a browser calling api.chess.com is subject to CORS, a
 * native client is not. It is also why the backend's proxy role disappears
 * rather than moving somewhere cheaper.
 */

import { Chess } from "chess.js";

export const BASE = "https://api.chess.com/pub";

export const USER_AGENT = "chess-analyzer/1.0";

export const DRAW_RESULTS = new Set([
  "agreed",
  "repetition",
  "stalemate",
  "insufficient",
  "50move",
  "timevsinsufficient",
]);

export class ChessComError extends Error {}

/**
 * @param {object} [http] Injected for tests; defaults to CapacitorHttp.
 */
export function createClient(http) {
  const request = async (url) => {
    const response = await http.get({
      url,
      headers: { "User-Agent": USER_AGENT, Accept: "application/json" },
      connectTimeout: 20_000,
      readTimeout: 20_000,
    });
    return response;
  };

  const getJson = async (url, { missingIsEmpty = false } = {}) => {
    const response = await request(url);
    if (response.status === 404) {
      if (missingIsEmpty) return null;
      throw new ChessComError(`Not found: ${url}`);
    }
    if (response.status >= 400) {
      throw new ChessComError(`Chess.com returned ${response.status} for ${url}`);
    }
    // The native layer parses JSON when the content type says so, and hands
    // back a string when it does not.
    return typeof response.data === "string" ? JSON.parse(response.data) : response.data;
  };

  return {
    async playerExists(username) {
      const response = await request(`${BASE}/player/${username.toLowerCase()}`);
      return response.status === 200;
    },

    /** Monthly archive URLs, oldest first. */
    async listArchives(username) {
      const data = await getJson(
        `${BASE}/player/${username.toLowerCase()}/games/archives`,
        { missingIsEmpty: true },
      );
      if (data === null) throw new ChessComError(`Chess.com user '${username}' not found`);
      return data.archives ?? [];
    },

    async fetchArchive(url) {
      const data = await getJson(url, { missingIsEmpty: true });
      return data === null ? [] : (data.games ?? []);
    },

    async fetchCurrentMonth(username) {
      const now = new Date();
      const month = String(now.getUTCMonth() + 1).padStart(2, "0");
      return this.fetchArchive(
        `${BASE}/player/${username.toLowerCase()}/games/${now.getUTCFullYear()}/${month}`,
      );
    },

    /** The last `months` archives, including the current one, oldest first. */
    async fetchRecentMonths(username, months) {
      const archives = await this.listArchives(username);
      const wanted = archives.slice(-Math.max(months, 1));
      const games = [];
      for (const url of wanted) games.push(...(await this.fetchArchive(url)));
      return games;
    },
  };
}

function pgnHeaders(pgn) {
  try {
    const game = new Chess();
    game.loadPgn(pgn);
    return game.getHeaders();
  } catch {
    return {}; // a malformed PGN must never kill a sync
  }
}

/**
 * The two kinds of game, and the whole of the rule that tells them apart.
 *
 * `training` is everything that does not measure the player's real strength:
 * games against the coach or a bot, and casual games. They still get analysed
 * and reviewed - the point of the split is that they do not sit in the same
 * average as rated play, where the result stood for something.
 *
 * The rule is `rated`, which Chess.com sends with every archived game and this
 * app has stored since the first version. It reads that one field, so it works
 * on a raw archive entry, on a normalised row and on a line of an old backup
 * alike - which is why the kind is filed at insert time rather than in
 * `normalizeGame`, whose output is compared field for field to a recording of
 * the Python backend and cannot gain one.
 *
 * Checked against a real archive of 227 games rather than assumed: the unrated
 * games were exactly the games whose PGN says `[Event "Play vs Coach"]`, all 33
 * of them against the same account, and nothing else in the archive was
 * unrated. The Event header is the more literal marker, and this is not it on
 * purpose - any coach or computer mode is unrated, so `rated` keeps working
 * when Chess.com adds another one under a different name. It errs only towards
 * being too broad, and too broad here means keeping a game that was not played
 * for a rating out of the average that stands for the player's rating.
 */
export const GAME_KINDS = ["rated", "training"];

export function gameKind(raw) {
  return raw?.rated === false ? "training" : "rated";
}

/** Chess.com puts the opening in ECOUrl, e.g. .../openings/Sicilian-Defense-Najdorf. */
function openingName(headers) {
  if (headers.Opening) return headers.Opening.slice(0, 160);
  const url = headers.ECOUrl;
  if (!url) return null;
  return url.replace(/\/+$/, "").split("/").pop().replaceAll("-", " ").slice(0, 160);
}

/**
 * Turn a Chess.com archive entry into a row for `games`.
 *
 * Returns null for anything we cannot analyse: variants, missing PGN, or a game
 * this player is not in.
 */
export function normalizeGame(raw, chessComUsername) {
  if (raw.rules !== "chess") return null;
  const pgn = raw.pgn;
  if (!pgn) return null;

  const me = chessComUsername.toLowerCase();
  const white = raw.white ?? {};
  const black = raw.black ?? {};

  let mine;
  let theirs;
  let color;
  if ((white.username ?? "").toLowerCase() === me) {
    [mine, theirs, color] = [white, black, "white"];
  } else if ((black.username ?? "").toLowerCase() === me) {
    [mine, theirs, color] = [black, white, "black"];
  } else {
    return null;
  }

  const rawResult = mine.result ?? "";
  let result;
  if (rawResult === "win") result = "win";
  else if (DRAW_RESULTS.has(rawResult)) result = "draw";
  else result = "loss";

  // Termination is whichever side's result code is not the generic "win".
  const termination = rawResult === "win" ? (theirs.result ?? null) : rawResult;

  const headers = pgnHeaders(pgn);
  const endTime = Number(raw.end_time ?? 0) || 0;

  // Only present once the game has been reviewed on Chess.com.
  const accuracies = raw.accuracies ?? {};

  return {
    chess_com_game_id: String(raw.uuid ?? raw.url ?? endTime),
    url: raw.url ?? null,
    pgn,
    user_color: color,
    user_rating: mine.rating ?? null,
    opponent_username: theirs.username ?? "?",
    opponent_rating: theirs.rating ?? null,
    result,
    termination,
    time_class: raw.time_class ?? null,
    time_control: raw.time_control ?? null,
    rated: Boolean(raw.rated ?? true),
    eco: headers.ECO || null,
    opening: openingName(headers),
    chess_com_accuracy: accuracies[color] ?? null,
    played_at: new Date(endTime * 1000).toISOString(),
    end_time: endTime,
  };
}
